web/src/games/bingo/board.tsx

```diff
@@ -0,0 +1,187 @@
+      this.props.moves.incrementCallRef(this._getPlayerID());
+    }
+  };
+
+  _numberClicked = (number: INumberState) => {
+    this.props.moves.playerClickedNumber(number, this._getPlayerID());
+  };
+
+  _shoutBingo = () => {
+    this.props.moves.playerShouted(this._getPlayerID());
+  };
+
+  _gameOverStatus = () => {
+    if (this.props.ctx.gameover.draw) {
+      return 'draw';
+    }
+    if (isOnlineGame(this.props.gameArgs)) {
+      if (this.props.ctx.gameover.winner === this.props.playerID) {
+        return 'you won';
+      } else {
+        return 'you lost';
+      }
+    } else {
+      return `Player ${this.props.ctx.gameover.winner} won`;
+    }
+  };
+
+  _renderFooter = () => {
+    const btnStyles = {
+      marginTop: '10px',
+      alignContent: 'center',
+    };
+
+    const stars = new Array(MAX_BINGO_CALLS)
+      .fill('⭐')
+      .map((s, idx) => (this.props.G.players[this._getPlayerID()].shoutCount > idx ? s : '✰'));
+
+    return (
+      <div style={{ display: 'flex' }}>
+        <Button
+          key="bi_call_table_btn"
+          variant="contained"
+          color="primary"
+          style={{
+            ...btnStyles,
+            justifyContent: 'left',
+          }}
+          onClick={() => this.setState({ showCallTable: !this.state.showCallTable })}
+        >
+          Table
+        </Button>
+        <Button
+          key="bi_shout_btn"
+          variant="contained"
+          color="primary"
+          style={{
+            ...btnStyles,
+            marginLeft: 'auto',
+            justifyContent: 'right',
+          }}
+          onClick={this._shoutBingo}
+        >
+          Bingo! {stars.join('')}
+        </Button>
+      </div>
+    );
+  };
+
+  _renderPlayComponents = (gameOver = false) => {
+    const { callQueue, callRef, players, timeRef, activePlayers } = this.props.G;
+    let playCard = (
+      <PlayCard
+        numbers={players[gameOver ? this.props.ctx.gameover.winner : this._getPlayerID()].numbers}
+        onNumberClicked={gameOver ? () => {} : this._numberClicked}
+      />
+    );
+    // if player has no Bingo! shouts left, then show special message
+    if (!gameOver && players[this._getPlayerID()].shoutCount <= 0) {
+      const msgLineHeight = 0.5;
+      playCard = (
+        <text
+          key={`bi_bingo_call_over_text`}
+          x="50%"
+          y={CALL_BOX_SIZE + 1.5}
+          fill="white"
+          textAnchor="middle"
+          fontSize={msgLineHeight * 0.8}
+        >
+          <tspan x="50%" dy="0">
+            Sorry, you used up
+          </tspan>
+          <tspan x="50%" dy={msgLineHeight}>
+            all your Bingo! calls ☹️
+          </tspan>
+          <tspan x="50%" dy={msgLineHeight * 2}>
+            Better luck next time !
+          </tspan>
+        </text>
+      );
+    }
+
+    // calculate backOff based on the list of activePlayers
+    // first person is 0 ms, then 500 ms, ..., and the rest will have max 500*length ms
```

Comment:
What happens if the first person internet delays 1s while the second person internet delays only 0.1s ? will this break the game?

Reply:
Well then the second person will fire before the first one and the game will continue. And thereafter the, second person will become the first person.

Reply:
I've just started reviewing this code, but it seems we're either missing a facility (async scheduling) or perhaps a better way to construct it out of existing infrastructure.  With the existing design I think that you should simplify the behavior so you're only ever in danger of sending one before a failover occurs:

in state: (current assignment: player1)
in state/config (assignment failover list (player1, 2, etc...)
in state (events with time displayed)

when the even schedule shows an expired event execute the update logic by current Assignment... else find failover assignment.

Reply:
I think my logic is a bit more evolved, it chooses a few candidates and puts them in a sequence and if all fail then it allows everyone to fire. I think what you are suggesting is that we have one candidate and if he fails, everyone tries. 
I made a lot of effort to implement this backoff mechanism, and would be a shame to downgrade.

Reply:
Another question I have is could a Timer be included in move logic (SetTimeout)?  This feels like a pub/sub problem. I like the authoritative state to be publisher as it keeps the chatter down on the network.

Reply:
> I think my logic is a bit more evolved, it chooses a few candidates and puts them in a sequence and if all fail then it allows everyone to fire. I think what you are suggesting is that we have one candidate and if he fails, everyone tries.
> I made a lot of effort to implement this backoff mechanism, and would be a shame to downgrade.

I'm not suggestion a downgrade, just a change in the failover selection being round-robin based upon a selection algorithm like you describe... I hear in your description a "shotgun mechanism" everybody tries at once. I'm still looking at the code where does this reside?

Reply:
Please try to follow variables BACKOFF_INTERVAL and G.activePlayers. As soon as a player fires timer, he comes to the top of the list of activePlayers and if he fails the next one to fire first comes to the top, simple ? 
And the number of players who can stay in the list is controlled by MAX_BACKOFF_CANDIDATES, currently 5 which makes the probabily nearly zero of every one firing at the same time.

Reply:
I just did a really crazy experiment, i had 6 windows open and tested, 
case 1: if all players leave, play resumes at the last player 
case 2: game continues even if only one player remains